KEP-2535: retarget alpha in 1.31 and update to readd the disk check (#4693)
kubernetes/enhancements: keps/sig-node/2535-ensure-secret-pulled-images/README.md

  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Test Plan](#test-plan)
      - [Prerequisite testing updates](#prerequisite-testing-updates)

>       - [Unit tests](#unit-tests)

Reviewer: Why these 2 sectionas are removed from the TOC? I can still see them below.

Author: they were duplicated afaict

Reviewer: I don't see it in the current content: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images

Author: true a rerun of update-toc fixed it, you're right :)

## Summary

Give the admin the ability to ensure pods that use an image are authorized to access that image. This will culminate in changes to the `IfNotPresent` and
`Never` pull policies, as the `Always` policy will go through an authentication check each time.

> When this feature is enabled, kubelet will check against an internal cache that maps images to credentials used. If the image has not been successfully pulled with the given credentials

Reviewer 2: All the information from here to the end of this section looks like it belongs to the proposal and/or implementation details sections.

Author: updated a bit to be less implementation detail heavy

since an unavailable registry will fail the pull image manifest check (with or without proper authentication.)

Thus, the motivation is to allow users to ensure the kubelet requires an image pull auth check for each new set of credentials,
regardless of whether the image is already present on the node.

### Goals

Modify the current behavior of images with an `IfNotPresent` and `Never` `ImagePullPolicy` enforced by the kubelet to
ensure the images pulled with a secret by the kubelet are authenticated by the CRI implementation. During the
EnsureImagesExist step the kubelet will require authentication of present images pulled with auth since boot.

Optimize to only force re-authentication for a pod container image when the
`ImagePullSecrets` used to pull the container image has not already been authenticated.
IOW if an image is pulled with authentication for a first pod, subsequent pods that have the same
authentication information should not need to re-authenticate, unless the kubelet's `pullImageSecretRecheckPeriod` has passed.

Images already present at boot or loaded externally to the kubelet or successfully

> pulled through the kubelet with no `ImagePullSecrets`/authentication required will

Reviewer 2: What about images pulled directly by the runtime with creds supplied to the runtime? Is that something to mention as a non goal initially?

Author: this is functionally the same as preloading the image, which the kubelet must treat as a non authorized pull as it doesn't have a cache entry for it

## Summary

We will add support in the kubelet for an admin to enable the ability to ensure an image that is already present on a node because

> a pod with `ImagePullSecrets` previously pulled it is reauthenticated when a new pod with different `ImagePullSecrets` attempts to use the same image,

Reviewer 3: ```suggestion
a pod with `ImagePullSecrets` previously pulled, it is reauthenticated when a new pod with different `ImagePullSecrets` attempts to use the same image,
```

when the `ImagePullPolicy` is `IfNotPresent`.

In other words: ensure the pull secrets are rechecked for each new set of credentials, and ensure a pod has access to those images.

> For the `Never` policy, the behavior also must change. Otherwise, a user who wishes to use the image of another pod could just use `Never` and hope

Author: note to a reviewer: this is a net-new aspect of this KEP, PTAL

> For the `Never` policy, the behavior also must change. Otherwise, a user who wishes to use the image of another pod could just use `Never` and hope

Reviewer 3: ```suggestion
For the `Never` policy, the behavior also must change, when opted in. Otherwise, a user who wishes to use the image of another pod could just use `Never` and hope
```

> another pod have pulled it. Functionally from a security standpoint, we must account for this.

Reviewer 3: ```suggestion
another pod has pulled it or that the image has been pre-loaded into the container runtime image cache. Functionally, from a security standpoint, we must account for this.
```

### Risks and Mitigations

- Image authentications with a registry may expire.
  - To mitigate expirations a timeout will be used to force re-authentication.
    This timeout will be configured as a kubelet configuration field `pullImageSecretRecheckPeriod`.
    This timeout feature will be implemented in alpha.

- Images can be "pre-loaded", or pulled behind the kubelet's back before it starts.
  In this case, the kubelet is not managing the credentials for these images.
  - To mitigate, metadata will be persisted across reboot. The kubelet will compare previously
    cached credentials against the images that exist. On a new image pull, the kubelet will use
    its saved cache and revalidate as necessary.
    In other words: even if the images are already cached, if new images are present that have not

>     previously been authenticated against a pods credentials, then the image will be revalidated.

Reviewer 3: Before beta, there should be further integration with the container runtimes to provide a validation/credentials process to support authentication/re-authentication of the pre-loaded images.

Author: do you mean a new CRI RPC other than PullImage?

Author: > we could do something different just for pre-loads and never

what's the difference between an auth check and a pull if the blob is already present on disk? We should do practically the same. 

> The ensure process being split as it is between kubelet, the container runtime, and the registries.. just doesn't make a lot of sense

fair, but I don't see this KEP as being the space to redefine this. I would suggest we scope this to fixing the existing cases and consider new integrations as a separate KEP

Reviewer 3: agreed.

garbage collected.. the list of images that required authentication in `kubelet`
will not be a source of truth for how all images were pulled that are in the
container runtime cache. To mitigate, images can be garbage collected at boot.
And we will persist ensure metadata across reboot of host, and restart
of kubelet, and possibly look at a way to add ensure metadata for images loaded
outside of kubelet. In beta we will add a switch to enable re-auth on boot for
admins seeking that instead of having to garbage collect where they do not use
or expect preloaded images since boot.

## Design Details

> Kubelet will track, in memory, a hash map for the credentials that were successfully used to pull an image.

Reviewer 2: nit: could we drop hash map in this update as it is an implementation detail? We could refer to it as `pulled image auth cache` or something similar.

[TestShouldPullImage link](https://github.com/kubernetes/kubernetes/pull/94899/files#diff-7297f08c72da9bf6479e80c03b45e24ea92ccb11c0031549e51b51f88a91f813R311-R438)

> PersistHashMeta() ** will be persisting SHA256 entries vs hash **

Reviewer 2: nit: Drop Hash from the name

Author: updated!
